### DISCUSSION TITLE
Filtering in a Redis steam on "Field" Field

### DISCUSSION THREAD

**Discussion by davutg** [redisson/redisson#5082] [Q&A] [answered]
As you know Redis streams are introduced after Redis v5. 
We're experimenting it in one of the project and we noticed that there's field and value sets available as JSON (at least redis-commander like consumers validate them for json format). I know technically Field/Value can be anything, but assumed json is the encouraged format and we used "Field" to keep some metadata about our objects. Now, we're willing to apply filters on that field if possible.

I would like to know if there's any way to apply filters on "Field". 

What we're doing currently is reading JSON strings to `RStream<String, String>` and casting manually to certain types. 
Also don't know why directly casting to object not working (This might be considered as another question)
codec= new StringCodec();  //current codec
codec = new TypedJsonJacksonCodec(ABC.class); //didn't work when we used and read `RStream<String,ABC>`

**Comment by mrniko** [comment, marked as answer]
Unable to reproduce any problem with TypedJsonJacksonCodec. Below is the code

```java
RStream<String, ABC> stream = redisson.getStream("stream", new TypedJsonJacksonCodec(RedissonStreamTest.ABC.class));
stream.createGroup("group");
stream.add(StreamAddArgs.entry("key", new ABC("value1")));

Map<StreamMessageId, Map<String, ABC>> s2 = stream.readGroup("group", "consumer2", StreamReadGroupArgs.neverDelivered());
```

**Reply by davutg** [reply]
Thanks @mrniko . Really enjoying using Redisson. It was our mistake using different codecs at pub/sub sides. Regarding the main concern do you have any suggestion ? We're putting some metadata of stream object to Redis Stream Field and don't know how to filter it on it. My colleague is trying to find a solution with https://github.com/RediSearch/JRediSearch . .. Basically while reading from the stream we wanna apply a query on 'field' (stream key) . 

Something like `stream.range(0, Long.MAX_VALUE, (msg) -> msg.get("name").equals("Nikita"));`

**Reply by davutg** [reply]
I believe even in REDIS base XRANGE command itself doesn't support such filtration. It would be very nice feature if it was supported. Now different consumers gonna read from the same stream and do filtration on the the read entries on client side. It doesn't look to me very nice practice, on the other hand using RediSearch and creating full text search is another drawback for the server end. If you have any suggestion for the best practice I appreciate.

**Reply by mrniko** [reply]
RediSearch has already been supported since 3.21.0 version 
https://github.com/redisson/redisson/wiki/9.-distributed-services/#96-redisearch-service

**Reply by atifmalik-com** [reply]
thanks @mrniko 

Mention above article is using RMap, RJsonBucket.

can we apply RediSearch with RStream?

**Reply by mrniko** [reply]
I'm afraid no. RediSearch isn't compatible with Stream.

**Reply by atifmalik-com** [reply]
@mrniko 
ok thank you for reply, can you tell us, if we are putting JSON keyField , for example,

String keyField = ""Name":/"abc/", "id":"233" " 
String value = ""val":/"433/", "test":"56444" "

stream.addAsync(StreamAddArgs.entries({keyField ,value }));

is there any way by filter KeyField for stream? I want to filter by keyField where ""Name":"abc""

**Reply by mrniko** [reply]
I'm afraid no.
